### PR TITLE
PersistentTuple implementation

### PR DESCRIPTION
**Pull request by catalin-manciu**
This adds the PersistentTuple class to the pmemobj module and implements tests for it.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pynvm/18)

### CODE REVIEW

**Comment by krzycz** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[nvm/pmemobj/tuple.py, line 29 at r1](https://reviewable.io:443/reviews/pmem/pynvm/18#-Kr171s_N3nBlZRWLTpk:-Kr171s_N3nBlZRWLTpl:b3oc34o) ([raw file](https://github.com/pmem/pynvm/blob/fc216e9737f4a6fa97699b74eb3aca9e32ae74d1/nvm/pmemobj/tuple.py#L29)):*
> ```Python
> 
>             ob = ffi.cast('PVarObject *', self._body)
>             mm.snapshot_range(ffi.addressof(ob, 'ob_size'),
> ```

Isn't the entire object already snapshotted in line 23? This seems to be unnecessary.

---

*[nvm/pmemobj/tuple.py, line 46 at r1](https://reviewable.io:443/reviews/pmem/pynvm/18#-Kr54FCawuHdrtQ0zZKX:-Kr54FCawuHdrtQ0zZKY:b8xw1ef) ([raw file](https://github.com/pmem/pynvm/blob/fc216e9737f4a6fa97699b74eb3aca9e32ae74d1/nvm/pmemobj/tuple.py#L46)):*
> ```Python
>             ob.ob_type = mm._get_type_code(PersistentTuple)
> 
>             self._body = ffi.cast('PTupleObject *', mm.direct(self._p_oid))
> ```

Is this a part of TX? Does it have to be? It's volatile, isn't it?
It's not a bug - just asking...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pynvm/18)*

**Comment by catalin-manciu** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[nvm/pmemobj/tuple.py, line 29 at r1](https://reviewable.io:443/reviews/pmem/pynvm/18#-Kr171s_N3nBlZRWLTpk:-Kr7kSzb_zy2w2oH8LtP:b-896fix) ([raw file](https://github.com/pmem/pynvm/blob/fc216e9737f4a6fa97699b74eb3aca9e32ae74d1/nvm/pmemobj/tuple.py#L29)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Isn't the entire object already snapshotted in line 23? This seems to be unnecessary.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pynvm/18)*

**Comment by catalin-manciu** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[nvm/pmemobj/tuple.py, line 46 at r1](https://reviewable.io:443/reviews/pmem/pynvm/18#-Kr54FCawuHdrtQ0zZKX:-Kr7l4MpP_2m-tf2aMO6:b-9g1w58) ([raw file](https://github.com/pmem/pynvm/blob/fc216e9737f4a6fa97699b74eb3aca9e32ae74d1/nvm/pmemobj/tuple.py#L46)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is this a part of TX? Does it have to be? It's volatile, isn't it?
It's not a bug - just asking...
</blockquote></details>

* `_body` is a Python wrapper over a non-volatile `PTupleObject` structure. 
* `ob` is a Python wrapper over the same structure but casted to a `PObject` (it only allows access to the first part of the `PTupleObject` structure which starts with such a `PObject`.
* `_body._ob_items` as well as `ob.ob_type` are persistent (since they're fields of the persistent `PTupleObject` structure) while `self` (the Python equivalent of **this**) is volatile so `self._body = ... ` is not part of the transaction but the code would be slightly less easier to follow if I move it outside.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pynvm/18)*

**Comment by krzycz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pynvm/18#-:-Kr9vOLisa8eLft2f_fn:bnfp4nl)*

File: nvm/pmemobj/tuple.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import collections
+import sys
+
+from .compat import recursive_repr, abc
+from .list import PersistentList
+from _pmem import ffi
+
+TUPLE_POBJPTR_ARRAY_TYPE_NUM = 50
```
Comment (catalin-manciu) [inline nvm/pmemobj/tuple.py]: This approach is used in dict, list, pool and now tuple. If there's a solution to this (Python doesn't have `typeid`, but you can still do `hash(type(var))` and get a hash based on the type of the variable `var`), it's going to be implemented in a separate pull request (since it's a widespread pattern).

Comment (tomaszkapela) [inline nvm/pmemobj/tuple.py]: Ok, so maybe I'll add one more commet. The hash should be stable, in the sense that it should give exactly the same, unique result between different _compilations_, runs, machines (or whatever else you might think of) for the same type. Otherwise you cannot rely on the type number to clearly identify your type (I've typed the word type a lot in the last two sentences 😄). That's the problem with C++, it doesn't guarantee that in the standard. The problem with hardcoding it is, that someone else is free to use that same number, rendering the feature useless.

Comment (bitdancer) [inline nvm/pmemobj/tuple.py]: Right, I hardcoded it originally with the intent of "fixing it later".  The python hash function is not guaranteed to be portable, even though for the most part it is.  One possibility is to use a hashlib function on the type name (like I did to get a portable hash in the dict implementation).
